src/Widgets/DiskBar.vala

```diff
@@ -0,0 +1,256 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Michael Aaron Murphy <michael@system76.com>
+ */
+
+public class Installer.DiskBar: Gtk.Grid {
+    public string disk_name { get; construct; }
+    public string disk_path { get; construct; }
+    public uint64 size { get; construct; }
+    public Gee.ArrayList<PartitionBar> partitions { get; construct; }
+
+    public Gtk.Box label;
+    private Gtk.Box bar;
+    private Gtk.ScrolledWindow legend;
+    private Gtk.Box legend_container;
+    private uint64 unused;
+
+    public DiskBar (
+        string model,
+        string path,
+        uint64 size,
+        Gee.ArrayList<PartitionBar> partitions
+    ) {
+        Object (
+            disk_name: model,
+            disk_path: path,
+            partitions: partitions,
+            size: size
+        );
+    }
+
+    construct {
+        unused = get_unused ();
+        generate_label ();
+        generate_bar ();
+        generate_legend ();
+
+        hexpand = true;
+        margin = 6;
+        get_style_context ().add_class ("storage-bar");
+        attach (label, 0, 1);
+        attach (legend, 1, 0);
+        attach (bar, 1, 1);
+
+        show_all ();
+    }
+
+    private uint64 get_unused () {
+        uint64 used = 0;
+        foreach (PartitionBar partition in partitions) {
+            used += partition.get_size ();
+        }
+
+        return size - (used * 512);
+    }
+
+    private void generate_legend () {
+        legend = new Gtk.ScrolledWindow (null, null);
+        legend.vscrollbar_policy = Gtk.PolicyType.NEVER;
+
+        legend_container = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 12);
+        legend_container.halign = Gtk.Align.CENTER;
+        legend_container.margin_bottom = 9;
+        legend.add (legend_container);
+
+        foreach (PartitionBar p in partitions) {
+            add_legend (p.path, p.get_size () * 512, Distinst.strfilesys (p.filesystem), p.vg, p.menu);
+        }
+
+        if (size / 100 < unused) {
+            add_legend ("unused", unused, "unused", null, null);
+        }
+    }
+
+    private void add_legend (string ppath, uint64 size, string fs, string? vg, Gtk.Popover? menu) {
+        var fill_round = new FillRound ();
+        fill_round.width_request = fill_round.height_request = 14;
+        fill_round.valign = Gtk.Align.CENTER;
+
+        var context = fill_round.get_style_context ();
+        context.add_class ("legend");
+        context.add_class (fs);
+
+        var format_size = GLib.format_size (size);
+
+        var info = new Gtk.Label (
+            (vg == null)
+                ? _("%s (%s)".printf (format_size, fs))
+                : _("%s (%s: <b>%s</b>)".printf (format_size, fs, vg))
```

Comment:
this won't work, it needs to be
_("%s (%s)").printf (format_size, fs)
_("%s (%s: <b>%s</b>)").printf (format_size, fs, vg)
otherwise it will try to translate the replaced string which won't be included in the translation catalog

Reply:
d'oh, thanks